### PR TITLE
Mask computed address in `Create2` and `Clones` libraries

### PR DESCRIPTION
**Pull request by byshape**
The `Create2` library has a `computeAddress` function and the `Clones` library has a `predictDeterministicAddress` function that return address as a result of the `keccak256`. Unmasked, this results in invalid value if used later in assembly code block.

Imagine that the `computeAddress` or `predictDeterministicAddress` function returns the address that is passed to another function that makes a call using that address as a parameter. If this assembly code, such as library, doesn’t clear the address, then this “address” will be invalid and the call will fail.

I suggest masking the result of the `keccak256` function to avoid possible problems with address usage.

#### PR Checklist

- [x] Tests
- [ ] Documentation
- [ ] Changeset entry (run `npx changeset add`)

### CODE REVIEW

File: contracts/mocks/ComputedAddressRemainderMock.sol
Comment (ernestognw) [inline contracts/mocks/ComputedAddressRemainderMock.sol:32]: We generally use the annotation syntax. No particular reason other than the Solidity docs stating that:

> In case you are developing a library that is meant to be compatible across multiple versions of Solidity, you can use a special comment to annotate an assembly block as memory-safe

```suggestion
        /// @solidity memory-safe-assembly
        assembly ("memory-safe") {
```

No big issue since it's a mock.

File: contracts/mocks/ComputedAddressRemainderMock.sol
Comment (ernestognw) [inline contracts/mocks/ComputedAddressRemainderMock.sol:1]: I'd split this mock in 2, one for Clones and the other for Create2, placing each file in their corresponding folder (e.g. /proxy and /utils)

Comment (ernestognw) [reply contracts/mocks/ComputedAddressRemainderMock.sol:1]: Tried to split them, but realized they make more sense together after all. Solving

**Review by ernestognw — DISMISSED** [review]
Thanks for this contribution @byshape!
Let's wait on @Amxx approval.

**Review by Amxx — DISMISSED** [review]

**Review by Amxx — APPROVED** [review]
I don't really see a need to check that in hardhat, as coverage doesn't report that kind of operations

I replaced the mock and hardhat tests with a very short/simple fuzzing test.

**Comment by ZumZoom** [comment]
Not sure about the visibility of the comment inside the resolved outdated changes so duplicating [it](https://github.com/OpenZeppelin/openzeppelin-contracts/pull/4941#discussion_r1528458911) here:

Is it still a good idea to stick with `/// @solidity memory-safe-assembly` annotations in favour of `assembly ("memory-safe")`?

**Comment by ernestognw** [comment]
> Is it still a good idea to stick with /// @solidity memory-safe-assembly annotations instead of assembly ("memory-safe")?
That's a good question.

I checked on the Solidity repository and found out this is where the assembly dialect was added.
https://github.com/ethereum/solidity/pull/12620

My understanding is that it was added in a non-breaking way, so we're fine keeping it as a NatSpec annotation but you're right in that we may prefer the dialect. I don't see any concrete benefit but I think we're open to discuss it further. I'm opening an issue.

**Review by ernestognw — APPROVED** [review]
I'm merging since the memory-safe annotation can be added in another PR and I don't want to block this one.

**Pull request merged**
